qcfractal/storage_sockets/db_queries.py

```diff
@@ -0,0 +1,193 @@
+from abc import ABC
+from qcelemental.util import msgpackext_dumps, msgpackext_loads
+from typing import List, Union
+from sqlalchemy.sql import text, bindparam
+
+
+class QueryBase(ABC):
+
+    # The name/alias used by the REST APIs to access this class
+    _class_name = None
+
+    # Mapping of the requested feature and the internal query method
+    _query_method_map = {}
+
+    def __init__(self, max_limit=10000):
+        self.max_limit = max_limit
+
+    def query(self, session, query_key, **kwargs):
+
+        if query_key not in self._query_method_map:
+            raise TypeError(f'Query type {query_key} is unimplemented for class {self._class_name}')
+
+        self.session = session
+
+        return getattr(self, self._query_method_map[query_key])(**kwargs)
+
+    def execute_query(self, sql_statement, with_keys=True, **kwargs):
+        """Execute sql statemet, apply limit, and return results as dict if needed"""
+
+         # TODO: check count first, way to iterate
+
+        # sql_statement += f' LIMIT {self.max_limit}'
+        result = self.session.execute(sql_statement, kwargs)
+        keys = result.keys()  # get keys before fetching
+        result = result.fetchall()
+
+        # create a list of dict with the keys and values of the results (instead of tuples)
+        if with_keys:
+            result = [dict(zip(keys, res)) for res in result]
+
+        return result
+
+    @staticmethod
+    def _raise_missing_attribute(cls, query_key, missing_attribute, amend_msg=''):
+        """Raises error for missinfg attribute in a message suitable for the REST user"""
+
+        raise AttributeError(f'To query {cls._class_name} for {query_key} '
+                             f'you must provide {missing_attribute}.')
+
+# ----------------------------------------------------------------------------
+
+class TorsionDriveQueries(QueryBase):
+
+    _class_name = 'torsiondrive'
+
+    _query_method_map = {
+        'initial_molecules' : '_get_initial_molecules',
+        'initial_molecules_ids' : '_get_initial_molecules_ids',
+        'final_molecules' : '_get_final_molecules',
+        'final_molecules_ids' : '_get_final_molecules_ids',
+        'return_results': '_get_return_results',  # TODO: maybe not used
+        'all_opt_results': '_get_all_opt_results',
+        'best_opt_results': '_get_best_opt_results'
+    }
+
+    def _get_initial_molecules_ids(self, torsion_id=None):
+
+        if torsion_id is None:
+            self._raise_missing_attribute('initial_molecules_ids', 'torsion drive id')
+
+        sql_statement = f"""
+                select initial_molecule from optimization_procedure as opt where opt.id in
+                (
+                    select opt_id from optimization_history where torsion_id = {torsion_id}
+                )  
+                order by opt.id
+        """
+
+        return self.execute_query(sql_statement, with_keys=False)
```

Comment:
We need to unpack and transform this to Pydantic objects at this layer as the socket layer does not do this transformation.

Reply:
I think it's better to keep them as dicts, and then move them back to the client where they transformed into Pydantic. This means one transformation only, instead of a second one before the REST, then serialize it again before sending to the client.

Reply:
We uniformly across the entire storage socket transform them to Pydantic objects before sending them to the client. This needs to be the standard to validate and we will have a switch to turn this off for production eventually.